migrations/move_site_icons.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ *
+ * phpBB Browser Push Notifications. An extension for the phpBB Forum Software package.
+ *
+ * @copyright (c) 2024, phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ */
+
+namespace phpbb\webpushnotifications\migrations;
+
+use phpbb\db\migration\container_aware_migration;
+use phpbb\filesystem\exception\filesystem_exception;
+use phpbb\filesystem\filesystem;
+
+class move_site_icons extends container_aware_migration
+{
+	private const NEW_ICON_DIR = 'images/site_icons';
+	private const OLD_ICON_DIR = 'images/icons';
+
+	/* @var filesystem $filesystem */
+	private $filesystem;
+
+	public function effectively_installed()
+	{
+		return $this->get_filesystem()->exists($this->container->getParameter('core.root_path') . self::NEW_ICON_DIR);
+	}
+
+	public static function depends_on()
+	{
+		return ['\phpbb\webpushnotifications\migrations\add_acp_pwa_configs'];
+	}
+
+	public function update_data(): array
+	{
+		return [
+			['custom', [[$this, 'configure_site_icons']]],
+		];
+	}
+
+	/**
+	 * Create a site_icons directory in the images directory (and copy existing PWA the icons there)
+	 *
+	 * @return void
+	 */
+	public function configure_site_icons()
+	{
+		// Cache frequently used services and values
+		$filesystem = $this->get_filesystem();
+		$root_path = $this->container->getParameter('core.root_path');
+		$user = $this->container->get('user');
+		$logger = $this->container->get('log');
+
+		// Prepare paths once
+		$new_icon_path = $root_path . self::NEW_ICON_DIR;
+		$old_icon_path = $root_path . self::OLD_ICON_DIR;
+
+		// Batch get config values
+		$icons = [
+			'small' => $this->config->offsetGet('pwa_icon_small'),
+			'large' => $this->config->offsetGet('pwa_icon_large')
+		];
+
+		try
+		{
+			// Create directory only if needed
+			if (!$filesystem->exists($new_icon_path))
+			{
+				$filesystem->mkdir($new_icon_path);
+			}
+
+			// Process icons
+			$copied = false;
+			foreach ($icons as $icon)
+			{
+				$old_file = $old_icon_path . '/' . $icon;
+				$new_file = $new_icon_path . '/' . $icon;
+
+				if (!empty($icon) && $filesystem->exists($old_file))
+				{
+					$filesystem->copy($old_file, $new_file);
+					$copied = true;
+				}
+			}
+
+			// Set up a log message result
+			$message = $copied ? 'LOG_WEBPUSH_ICON_COPY_SUCCESS' : 'LOG_WEBPUSH_ICON_DIR_SUCCESS';
+			$result = [$message => [self::NEW_ICON_DIR]];
+		}
+		catch (filesystem_exception $e)
```

Comment:
Not sure how try/catch will work within migration. Feels like it'll throw uncaught exception on failure which will stop extension installation and break things on user end.

Reply:
That's why there's the catch. It'll gracefully catch any of the exceptions the filesystem throws and log the event. Doesn't stop installation.